<a href="https://colab.research.google.com/github/Chau-Nguyen-Developer/HeHeHaHa_DataRoyale/blob/main/Chau_RoyaleClash.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install gdown

In [ ]:
import gdown

In [ ]:
file_id = "1sgMvwiuZNyBt86JD3FRVZa5YRi6DpbBj"

In [ ]:
url = f"https://drive.google.com/uc?id={file_id}"

@everyone DATASET IS HERE WOOT: https://drive.google.com/file/d/1sgMvwiuZNyBt86JD3FRVZa5YRi6DpbBj/view?usp=drive_link 🎊

In [ ]:
# Download the file
gdown.download(url, "dataset.csv", quiet=False)
print("\nFinished downloading dataset.")


Downloading...
From (original): https://drive.google.com/uc?id=1sgMvwiuZNyBt86JD3FRVZa5YRi6DpbBj
From (redirected): https://drive.google.com/uc?id=1sgMvwiuZNyBt86JD3FRVZa5YRi6DpbBj&confirm=t&uuid=5852d91f-1d11-4e7f-ac57-6913579d13b6
To: /content/dataset.csv
100%|██████████| 9.87G/9.87G [01:26<00:00, 114MB/s]



Finished downloading dataset.


In [ ]:
# # Load data using its path
# import pandas as pd

# path = "/content/dataset.csv"
# data = pd.read_csv(path)
# print("\nFinish loading.")

In [ ]:
import pandas as pd
data = pd.read_csv("dataset.csv", nrows=800_000)
print("Done loading.")

Done loading.


In [ ]:
# Check the structure
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800000 entries, 0 to 799999
Data columns (total 74 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   Unnamed: 0                      800000 non-null  int64  
 1   battleTime                      800000 non-null  object 
 2   arena.id                        800000 non-null  float64
 3   gameMode.id                     800000 non-null  float64
 4   average.startingTrophies        800000 non-null  float64
 5   winner.tag                      800000 non-null  object 
 6   winner.startingTrophies         800000 non-null  float64
 7   winner.trophyChange             800000 non-null  float64
 8   winner.crowns                   800000 non-null  float64
 9   winner.kingTowerHitPoints       800000 non-null  float64
 10  winner.princessTowersHitPoints  784727 non-null  object 
 11  winner.clan.tag                 794897 non-null  object 
 12  winner.clan.badg

Columns of winner.princessTowersHitPoints, winner.clan.tag, winner.clan. badgeId, winner.princessTowersHitPoints, winner.clan.tag, winner.clan.badgeId are only columns that has NULL entries.

In [ ]:
# Descriptive Statistics
display(data.describe())

Unnamed: 0      arena.id   gameMode.id  average.startingTrophies  \
count  800000.000000  8.000000e+05  8.000000e+05             800000.000000   
mean   399999.500000  5.400005e+07  7.200011e+07               4816.478458   
std    230940.252015  7.331788e+00  9.727730e+01                631.484620   
min         0.000000  5.400000e+07  7.200001e+07                 26.000000   
25%    199999.750000  5.400005e+07  7.200001e+07               4538.000000   
50%    399999.500000  5.400005e+07  7.200020e+07               4818.500000   
75%    599999.250000  5.400005e+07  7.200020e+07               5145.500000   
max    799999.000000  5.400005e+07  7.200020e+07               8184.000000   

       winner.startingTrophies  winner.trophyChange  winner.crowns  \
count            800000.000000        800000.000000  800000.000000   
mean               4816.969311            29.449180       1.769619   
std                 631.688339             1.502183       0.844154   
min                  26.000000             1.000000       1.000000   
25%                4538.000000            29.000000       1.000000   
50%                4819.000000            30.000000       2.000000   
75%                5146.000000            30.000000       3.000000   
max                8193.000000            58.000000       3.000000   

       winner.kingTowerHitPoints  winner.clan.badgeId  loser.startingTrophies  \
count              800000.000000         7.948970e+05           800000.000000   
mean                 4957.712090         1.600008e+07             4815.987605   
std                  1127.140168         5.849747e+01              631.396175   
min                     1.000000         1.600000e+07               26.000000   
25%                  4392.000000         1.600003e+07             4537.000000   
50%                  5304.000000         1.600007e+07             4818.000000   
75%                  5832.000000         1.600014e+07             5145.000000   
max                  6999.000000         1.600018e+07             8185.000000   

       ...  loser.card8.level  loser.totalcard.level  loser.troop.count  \
count  ...      800000.000000          800000.000000      800000.000000   
mean   ...          11.901200              95.337126           5.617351   
std    ...           1.701335              13.061369           1.104258   
min    ...           1.000000               8.000000           0.000000   
25%    ...          11.000000              90.000000           5.000000   
50%    ...          13.000000             102.000000           6.000000   
75%    ...          13.000000             104.000000           6.000000   
max    ...          13.000000             104.000000           8.000000   

       loser.structure.count  loser.spell.count  loser.common.count  \
count          800000.000000      800000.000000       800000.000000   
mean                0.452316           1.930333            2.085237   
std                 0.613394           0.887981            1.312746   
min                 0.000000           0.000000            0.000000   
25%                 0.000000           1.000000            1.000000   
50%                 0.000000           2.000000            2.000000   
75%                 1.000000           2.000000            3.000000   
max                 8.000000           8.000000            8.000000   

       loser.rare.count  loser.epic.count  loser.legendary.count  \
count     800000.000000     800000.000000          800000.000000   
mean           2.080706          2.219899               1.614157   
std            1.279775          1.390577               1.220508   
min            0.000000          0.000000               0.000000   
25%            1.000000          1.000000               1.000000   
50%            2.000000          2.000000               1.000000   
75%            3.000000          3.000000               2.000000   
max            8.000000          8.000000               8.000000   

       loser.elix

In [ ]:
# Check some of the entries in the head
data.head()

Unnamed: 0                 battleTime    arena.id  gameMode.id  \
0           0  2020-12-07 07:00:00+00:00  54000049.0   72000201.0   
1           1  2020-12-07 07:00:00+00:00  54000049.0   72000201.0   
2           2  2020-12-07 07:00:02+00:00  54000049.0   72000201.0   
3           3  2020-12-07 07:00:03+00:00  54000049.0   72000201.0   
4           4  2020-12-07 07:00:06+00:00  54000049.0   72000201.0   

   average.startingTrophies  winner.tag  winner.startingTrophies  \
0                    6590.0  #28RR8PJP0                   6581.0   
1                    5582.5   #YV9VQUVP                   5592.0   
2                    5684.0  #LPR2G0Q9L                   5678.0   
3                    6031.0  #2GL899VCJ                   6035.0   
4                    5140.0  #9Y2YJPGG2                   5140.0   

   winner.trophyChange  winner.crowns  winner.kingTowerHitPoints  ...  \
0                 31.0            2.0                     4768.0  ...   
1                 28.0            3.0                     2014.0  ...   
2                 31.0            3.0                     5304.0  ...   
3                 29.0            2.0                     3368.0  ...   
4                 30.0            3.0                     1507.0  ...   

                                    loser.cards.list loser.totalcard.level  \
0  [26000000, 26000026, 26000030, 26000041, 27000...                   104   
1  [26000000, 26000003, 26000007, 26000011, 26000...                   104   
2  [26000011, 26000026, 26000030, 26000041, 27000...                   103   
3  [26000032, 26000040, 26000041, 26000049, 26000...                   104   
4  [26000012, 26000024, 26000045, 26000056, 26000...                    93   

   loser.troop.count loser.structure.count  loser.spell.count  \
0                  4                     1                  3   
1                  6                     0                  2   
2                  4                     1                  3   
3                  6                     1                  1   
4                  5                     1                  2   

   loser.common.count  loser.rare.count  loser.epic.count  \
0                   4                 1                 1   
1                   2                 3                 3   
2                   3                 2                 2   
3                   3                 2                 1   
4                   3                 1                 4   

  loser.legendary.count  loser.elixir.average  
0                     2                 3.125  
1                     0                 4.125  
2                     1                 2.875  
3                     2                 3.375  
4                     0                 3.875  

[5 rows x 74 columns]

In [10]:
data["winner.elixir.average"]

0         3.625
1         4.125
2         3.875
3         3.875
4         4.125
          ...  
799995    3.875
799996    4.125
799997    3.750
799998    4.125
799999    2.625
Name: winner.elixir.average, Length: 800000, dtype: float64

In [13]:
# See all the name of columns
data.columns

Index(['Unnamed: 0', 'battleTime', 'arena.id', 'gameMode.id',
       'average.startingTrophies', 'winner.tag', 'winner.startingTrophies',
       'winner.trophyChange', 'winner.crowns', 'winner.kingTowerHitPoints',
       'winner.princessTowersHitPoints', 'winner.clan.tag',
       'winner.clan.badgeId', 'loser.tag', 'loser.startingTrophies',
       'loser.trophyChange', 'loser.crowns', 'loser.kingTowerHitPoints',
       'loser.clan.tag', 'loser.clan.badgeId', 'loser.princessTowersHitPoints',
       'tournamentTag', 'winner.card1.id', 'winner.card1.level',
       'winner.card2.id', 'winner.card2.level', 'winner.card3.id',
       'winner.card3.level', 'winner.card4.id', 'winner.card4.level',
       'winner.card5.id', 'winner.card5.level', 'winner.card6.id',
       'winner.card6.level', 'winner.card7.id', 'winner.card7.level',
       'winner.card8.id', 'winner.card8.level', 'winner.cards.list',
       'winner.totalcard.level', 'winner.troop.count',
       'winner.structure.count', 'winner.

In [26]:
data.head(1)

Unnamed: 0                 battleTime    arena.id  gameMode.id  \
0           0  2020-12-07 07:00:00+00:00  54000049.0   72000201.0   

   average.startingTrophies  winner.tag  winner.startingTrophies  \
0                    6590.0  #28RR8PJP0                   6581.0   

   winner.trophyChange  winner.crowns  winner.kingTowerHitPoints  ...  \
0                 31.0            2.0                     4768.0  ...   

                                    loser.cards.list loser.totalcard.level  \
0  [26000000, 26000026, 26000030, 26000041, 27000...                   104   

   loser.troop.count loser.structure.count  loser.spell.count  \
0                  4                     1                  3   

   loser.common.count  loser.rare.count  loser.epic.count  \
0                   4                 1                 1   

  loser.legendary.count  loser.elixir.average  
0                     2                 3.125  

[1 rows x 74 columns]

In [27]:
data.head(1)["winner.trophyChange"]

0    31.0
Name: winner.trophyChange, dtype: float64

In [17]:
data["winner.cards.list"][0]

'[26000036, 26000043, 26000044, 26000050, 26000054, 26000062, 28000015, 28000016]'

In [28]:
data.head(1)["winner.cards.list"]

0    [26000036, 26000043, 26000044, 26000050, 26000...
Name: winner.cards.list, dtype: object

In [25]:
data["loser.cards.list"][0]

'[26000000, 26000026, 26000030, 26000041, 27000006, 28000003, 28000004, 28000011]'

In [23]:
data.head(1)["winner.tag"]

0    #28RR8PJP0
Name: winner.tag, dtype: object

In [24]:
data.head(1)["loser.tag"]

0    #9UU8P008P
Name: loser.tag, dtype: object

In [31]:
data.head(1)["winner.kingTowerHitPoints"]

0    4768.0
Name: winner.kingTowerHitPoints, dtype: float64

In [32]:
data.head(1)["winner.card1.level"]

0    13
Name: winner.card1.level, dtype: int64

In [33]:
data.head(1)["winner.crowns"]

0    2.0
Name: winner.crowns, dtype: float64

In [34]:
data.head(1)["loser.crowns"]

0    1.0
Name: loser.crowns, dtype: float64

In [35]:
data.head(1)["winner.clan.tag"]

0    #YYL2GJUC
Name: winner.clan.tag, dtype: object

In [36]:
data.head(1)["loser.clan.tag"]

0    #P2PUU9JV
Name: loser.clan.tag, dtype: object

### Clan Performance
Metrics for each clan:
- Win rate
- Average crown difference (winner.crowns - loser.crowns)
- Average trophy change (trophy gains/losses per match)
- Net trophies (sum of trophy changes across all members of that clan)
- Stability (lower standard deviation of trophy change = more consistent)

In [37]:
data.columns


Index(['Unnamed: 0', 'battleTime', 'arena.id', 'gameMode.id',
       'average.startingTrophies', 'winner.tag', 'winner.startingTrophies',
       'winner.trophyChange', 'winner.crowns', 'winner.kingTowerHitPoints',
       'winner.princessTowersHitPoints', 'winner.clan.tag',
       'winner.clan.badgeId', 'loser.tag', 'loser.startingTrophies',
       'loser.trophyChange', 'loser.crowns', 'loser.kingTowerHitPoints',
       'loser.clan.tag', 'loser.clan.badgeId', 'loser.princessTowersHitPoints',
       'tournamentTag', 'winner.card1.id', 'winner.card1.level',
       'winner.card2.id', 'winner.card2.level', 'winner.card3.id',
       'winner.card3.level', 'winner.card4.id', 'winner.card4.level',
       'winner.card5.id', 'winner.card5.level', 'winner.card6.id',
       'winner.card6.level', 'winner.card7.id', 'winner.card7.level',
       'winner.card8.id', 'winner.card8.level', 'winner.cards.list',
       'winner.totalcard.level', 'winner.troop.count',
       'winner.structure.count', 'winner.

Get rid of column "Tournament Tag"as all entries has NaN value for tournament tag.

In [41]:
data["tournamentTag"].isna().sum()


np.int64(800000)

In [42]:
data["gameMode.id"]

0         72000201.0
1         72000201.0
2         72000201.0
3         72000201.0
4         72000201.0
             ...    
799995    72000201.0
799996    72000201.0
799997    72000201.0
799998    72000201.0
799999    72000201.0
Name: gameMode.id, Length: 800000, dtype: float64

In [43]:
data.head(1)["winner.clan.tag"]

0    #YYL2GJUC
Name: winner.clan.tag, dtype: object

In [44]:
data.head(1)["winner.clan.badgeId"]

0    16000111.0
Name: winner.clan.badgeId, dtype: float64

In [45]:
data.head(1)["winner.cards.list"]

0    [26000036, 26000043, 26000044, 26000050, 26000...
Name: winner.cards.list, dtype: object

In [46]:
data.head(1)["winner.totalcard.level"]

0    104
Name: winner.totalcard.level, dtype: int64

In [47]:
# Filter out the columns that I need to compare clan performance
cols = [ 'battleTime',
         'average.startingTrophies',
         'winner.tag',
         'winner.clan.tag',
         'winner.clan.badgeId',
         'winner.startingTrophies',
         'winner.trophyChange',
         'winner.crowns',
         'winner.cards.list',
         'winner.totalcard.level',
         'winner.elixir.average',
         'loser.tag',
         'loser.clan.tag',
         'loser.clan.badgeId',
         'loser.startingTrophies',
         'loser.trophyChange',
         'loser.crowns',
         'loser.cards.list',
         'loser.totalcard.level',
         'loser.elixir.average']




In [52]:
clan_df = data[cols]

In [53]:
clan_df.columns

Index(['battleTime', 'average.startingTrophies', 'winner.tag',
       'winner.startingTrophies', 'winner.trophyChange', 'winner.crowns',
       'winner.clan.tag', 'winner.clan.badgeId', 'winner.cards.list',
       'winner.totalcard.level', 'winner.elixir.average', 'loser.tag',
       'loser.startingTrophies', 'loser.trophyChange', 'loser.crowns',
       'loser.clan.tag', 'loser.clan.badgeId', 'loser.cards.list',
       'loser.totalcard.level', 'loser.elixir.average'],
      dtype='object')

### Reshape one row (include winner and loser's info) per match to per player per match

Separate winners and losers and put them into separate rows. One player = One row.

In [56]:
winners = pd.DataFrame({
    "playerTag": data["winner.tag"],
    "battleTime": data["battleTime"],
    "clanTag": data["winner.clan.tag"],
    "playerStartTrophies": data["winner.startingTrophies"],
    "opponentStartTrophies": data["loser.startingTrophies"],
    "trophyChange": data["winner.trophyChange"],
    "crownsGained": data["winner.crowns"],
    "crownsLost": data["loser.crowns"],
    "result": 1 # Represent winning
})

In [57]:
losers = pd.DataFrame({
    "playerTag": data["loser.tag"],
    "battleTime": data["battleTime"],
    "clanTag": data["loser.clan.tag"],
    "playerStartTrophies": data["loser.startingTrophies"],
    "opponentStartTrophies": data["winner.startingTrophies"],
    "trophyChange": data["loser.trophyChange"],
    "crownsGained": data["loser.crowns"],
    "crownsLost": data["winner.crowns"],
    "result": 0 # Represent losing
})

In [58]:
perPlayer = pd.concat([winners, losers], ignore_index=True)
# Generate fresh indices starting from 0 with ignore_index = True
print(type(perPlayer))

<class 'pandas.core.frame.DataFrame'>


In [60]:
perPlayer.head(5)

playerTag                 battleTime    clanTag  playerStartTrophies  \
0  #28RR8PJP0  2020-12-07 07:00:00+00:00  #YYL2GJUC               6581.0   
1   #YV9VQUVP  2020-12-07 07:00:00+00:00    #QJ2PJ8               5592.0   
2  #LPR2G0Q9L  2020-12-07 07:00:02+00:00    #LC8QY8               5678.0   
3  #2GL899VCJ  2020-12-07 07:00:03+00:00  #2LPULRC2               6035.0   
4  #9Y2YJPGG2  2020-12-07 07:00:06+00:00   #8C899GY               5140.0   

   opponentStartTrophies  trophyChange  crownsGained  crownsLost  result  
0                 6599.0          31.0           2.0         1.0       1  
1                 5573.0          28.0           3.0         1.0       1  
2                 5690.0          31.0           3.0         0.0       1  
3                 6027.0          29.0           2.0         1.0       1  
4                 5140.0          30.0           3.0         1.0       1

In [62]:
# Check to see if any clanTage is Nan
perPlayer["clanTag"].isna().sum()

np.int64(63656)

There are 63,656 players who do not have clan tag. Will drop them in this analysis.

In [63]:
perPlayer = perPlayer.dropna(subset=["clanTag"])

In [64]:
import numpy as np

"size" --> count how many matches that clan play (aka how many rows per clan)  
"sum" --> count how many wins that clan wins (add up the 1's in the result)

Syntax for .agg()
newColumnName = (existing_column, aggregation_function)

In [68]:
# test if any negative values in "trophy change"column
hasNeg = (perPlayer["trophyChange"] < 0).any()
print(hasNeg)

True


In [72]:
negTrophyChange = perPlayer[perPlayer["trophyChange"] < 0]["trophyChange"]
negTrophyChange

800000    -31.0
800001    -28.0
800002    -31.0
800003    -29.0
800004    -30.0
           ... 
1599994   -27.0
1599995   -32.0
1599996   -26.0
1599997   -30.0
1599999   -29.0
Name: trophyChange, Length: 741447, dtype: float64

In [73]:
# Group each player into their clan tags.
perClanAggregates = perPlayer.groupby("clanTag").agg(
    matches=("result", "size"),
    uniquePlayers=("playerTag", pd.Series.nunique),

    wins=("result", "sum"),
    winRate=("result", "mean"),

    clanCrownGainedTotal = ("crownsGained", "sum"),
    clanCronwGainedAVG = ("crownsGained", "mean"),

    clanTrophyTotal = ("trophyChange", "sum"),
    clanTrophyChange = ("trophyChange", "mean"),

    opponentStartTrophiesAVG = ("opponentStartTrophies", "mean")
).reset_index()

In [74]:
perPlayer.head(5)

playerTag                 battleTime    clanTag  playerStartTrophies  \
0  #28RR8PJP0  2020-12-07 07:00:00+00:00  #YYL2GJUC               6581.0   
1   #YV9VQUVP  2020-12-07 07:00:00+00:00    #QJ2PJ8               5592.0   
2  #LPR2G0Q9L  2020-12-07 07:00:02+00:00    #LC8QY8               5678.0   
3  #2GL899VCJ  2020-12-07 07:00:03+00:00  #2LPULRC2               6035.0   
4  #9Y2YJPGG2  2020-12-07 07:00:06+00:00   #8C899GY               5140.0   

   opponentStartTrophies  trophyChange  crownsGained  crownsLost  result  
0                 6599.0          31.0           2.0         1.0       1  
1                 5573.0          28.0           3.0         1.0       1  
2                 5690.0          31.0           3.0         0.0       1  
3                 6027.0          29.0           2.0         1.0       1  
4                 5140.0          30.0           3.0         1.0       1

In [77]:
perClanAggregates["clanTag"]

0         #200000JU
1          #20000VC
2          #200020C
3          #200020P
4         #200028G9
            ...    
316026    #YYYYUR2J
316027    #YYYYUR80
316028    #YYYYVPLG
316029    #YYYYVQ0R
316030    #YYYYVRLV
Name: clanTag, Length: 316031, dtype: object

### Number of unique clans: 316,031

In [78]:
perClanAggregates

clanTag  matches  uniquePlayers  wins   winRate  \
0       #200000JU        1              1     0  0.000000   
1        #20000VC        5              4     4  0.800000   
2        #200020C        2              2     0  0.000000   
3        #200020P        7              5     4  0.571429   
4       #200028G9        9              6     4  0.444444   
...           ...      ...            ...   ...       ...   
316026  #YYYYUR2J        3              3     1  0.333333   
316027  #YYYYUR80        2              2     0  0.000000   
316028  #YYYYVPLG      543             23   543  1.000000   
316029  #YYYYVQ0R        2              2     0  0.000000   
316030  #YYYYVRLV        1              1     0  0.000000   

        clanCrownGainedTotal  clanCronwGainedAVG  clanTrophyTotal  \
0                        0.0            0.000000            -30.0   
1                        9.0            1.800000             89.0   
2                        2.0            1.000000            -57.0   
3                        9.0            1.285714             34.0   
4                        9.0            1.000000            -20.0   
...                      ...                 ...              ...   
316026                   4.0            1.333333            -33.0   
316027                   0.0            0.000000            -65.0   
316028                 976.0            1.797422          15964.0   
316029                   0.0            0.000000            -59.0   
316030                   1.0            1.000000            -30.0   

        clanTrophyChange  opponentStartTrophiesAVG  
0             -30.000000               4715.000000  
1              17.800000               4602.000000  
2             -28.500000               4766.500000  
3               4.857143               5006.857143  
4              -2.222222               4804.111111  
...                  ...                       ...  
316026        -11.000000               4498.333333  
316027        -32.500000               4734.000000  
316028         29.399632               4485.064457  
316029        -29.500000               4667.000000  
316030        -30.000000               3029.000000  

[316031 rows x 10 columns]